### ISSUE TITLE
Table cell widths ignored?

### ISSUE THREAD

**Issue by leei** [urvin-compliance/caracal#123]
Table cell widths seem to be accepted but ignored in modern versions of Word (16.25 for Mac and current Word on Windows 10.  When I open a docx with Pages or Open Office, it seems to be respecting the twips-based cell widths, but in Word it seems to default to just auto-width even when twips are given.

I'm trying to get percentage width columns so I'm effectively going with:

container_width = docx.page_width - docx.page_margin_left - docx.page_margin_right
col_width = container_width / 3

and then 

cell = Caracal::Core::Models::TableCellModel.new do |cell|
  cell.width col_width
  cell.h4 "Direct Reports"
end

Any hints? Do I need to do the calculation differently?

**Comment by jdugan** [comment]
I gather that this is somewhere between code and pseudocode, but before we dive in too deep, your example above uses the variable `thirds` to set the width, but you call that variable `col_width` in the code above it.

Is that a function of how you translated/presented the code in the issue description, or is it possible you are passing a `nil` or similar into the `width` method?  Which some versions might interpret to mean divide the widths evenly and others to mean autofit?

**Comment by leei** [comment]
Yes, just corrected the example. Ideally, I'd like to be able to specify a pct width without needing to calculate...

**Comment by martinsp** [comment]
This pull request might be related: #115 . It allows to set column widths, but does not support percentages.

Without this patch caracal "auto-calculates" column widths from the first row. If you specify cell width in a row other than first row, it will be ignored.

**Comment by leei** [comment]
@martinsp Your pull request is probably related. The issue with my table is a header row with colspan: 2 or 3. That seems to be the precipitating problem. I'm going to try a few things in caracal-example to test your theory.

**Comment by leei** [comment]
Here's a replicable example (inserted into caracal-example doc):

```
docx.table [['Only Header'], row2.dup, row3.dup, row4.dup], border_size: 4 do
  cell_style rows[0],    bold: true,   background: '3366cc', color: 'ffffff', colspan: 2
  cell_style rows[-1],   bold: true,   background: 'dddddd'
  cell_style cells[5],   italic: true, color: 'cc0000'
  cell_style cols[0],    width: 6000
  cell_style cols[1],    width: 3000
  cell_style cells,      size: 18, margins: { top: 100, bottom: 0, left: 100, right: 100 }
end
```

This generates a table with a header row and then two "equal width" columns... Yikes!

**Comment by dlauer** [comment]
Closing — the root cause identified in this thread is fixed. As @leei narrowed it down, the trigger was a header row with `colspan`: the table grid was derived from the first row, so one merged cell flattened every column to the same width.

#115 (now merged) adds `column_widths`, which sets the grid explicitly:

```ruby
docx.table [['Spans two columns'], ['Cell 1', 'Cell 2']] do
  cell_style rows[0][0], colspan: 2
  column_widths [2400, 1200]
end
```

Verified: that renders a 2400/1200 grid, where the same table without `column_widths` renders 4680/4680. Thanks @martinsp for the patch and for updating it.

Percentage-based widths (`w:type="pct"`) are a separate feature — please open a new issue if you'd still like those.
